desmod/pool.py

```diff
@@ -0,0 +1,197 @@
+empty or full. Users can put or get items in the pool with a certain amount as
+a parameter.
+"""
+
+from simpy import Event
+from simpy.core import BoundClass
+
+
+class PoolPutEvent(Event):
+    def __init__(self, pool, amount=1):
+        super(PoolPutEvent, self).__init__(pool.env)
+        self.pool = pool
+        self.amount = amount
+        self.callbacks.append(pool._trigger_get)
+        pool._putters.append(self)
+        pool._trigger_put()
+
+    def cancel(self):
+        if not self.triggered:
+            self.pool._putters.remove(self)
+            self.callbacks = None
+
+
+class PoolGetEvent(Event):
+    def __init__(self, pool, amount=1):
+        super(PoolGetEvent, self).__init__(pool.env)
+        self.pool = pool
+        self.amount = amount
+        self.callbacks.append(pool._trigger_put)
+        pool._getters.append(self)
+        pool._trigger_get()
+
+    def cancel(self):
+        if not self.triggered:
+            self.pool._getters.remove(self)
+            self.callbacks = None
+
+
+class PoolWhenNewEvent(Event):
+    def __init__(self, pool):
+        super(PoolWhenNewEvent, self).__init__(pool.env)
+        self.pool = pool
+        pool._new_waiters.append(self)
+
+    def cancel(self):
+        if not self.triggered:
+            self.pool._new_waiters.remove(self)
+            self.callbacks = None
+
+
+class PoolWhenAnyEvent(Event):
+    def __init__(self, pool):
+        super(PoolWhenAnyEvent, self).__init__(pool.env)
+        self.pool = pool
+        pool._any_waiters.append(self)
+        pool._trigger_when_any()
+
+    def cancel(self):
+        if not self.triggered:
+            self.pool._any_waiters.remove(self)
+            self.callbacks = None
+
+
+class PoolWhenFullEvent(Event):
+    def __init__(self, pool):
+        super(PoolWhenFullEvent, self).__init__(pool.env)
+        self.pool = pool
+        pool._full_waiters.append(self)
+        pool._trigger_when_full()
+
+    def cancel(self):
+        if not self.triggered:
+            self.pool._full_waiters.remove(self)
+            self.callbacks = None
+
+
+class Pool(object):
+    """Simulation pool of discrete or continuous resources.
+
+    `Pool` is similar to :class:`simpy.resources.Container`.
+    It provides a simulation-aware container for managing a shared pool of
+    resources. The resources can be either discrete objects (like apples) or
+    continuous (like water).
+
+    Resources are added and removed using :meth:`put()` and :meth:`get()`.
+
+    :param env: Simulation environment.
+    :param capacity: Capacity of the pool; infinite by default.
+    :param hard_cap:
+        If specified, the pool overflows when the `capacity` is reached.
+    :param init_level: Initial level of the pool.
+    :param name: Optional name to associate with the queue.
+
+    """
+
+    def __init__(self, env, capacity=float('inf'), hard_cap=False,
+                 init_level=0, name=None):
```

Comment:
I suggest renaming `init_level` to `init` for compatibility with `simpy.Container`.

For the same reason, I also suggest the parameter order be `(env, capacity, init, hard_cap, name)` so that `Container`'s specifying `init` positionally can be trivially replaced with `Pool`.